src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java

```diff
@@ -162,21 +173,20 @@ public void addConfiguration(InstanceConfiguration config) {
                     public Node call() throws Exception {
                         long startTime = System.currentTimeMillis();
                         LOGGER.log(Level.INFO, String.format("Waiting %dms for node %s to connect", config.getLaunchTimeoutMillis(), node.getNodeName()));
-                        while ((System.currentTimeMillis() - startTime) < config.getLaunchTimeoutMillis()) {
-                            LOGGER.log(Level.INFO, String.format("%dms elapsed waiting for node %s to connect", System.currentTimeMillis() - startTime, node.getNodeName()));
-                            Computer c;
-                            try {
-                                c = node.toComputer();
-                                if (c != null) {
-                                    c.connect(false).get();
-                                }
-                            } catch (Exception e) {
-                                LOGGER.log(Level.WARNING, String.format("Exception waiting for node %s to connect", node.getNodeName()), e);
+                        try {
+                            Computer c = node.toComputer();
+                            if (c != null) {
+                                c.connect(false).get(config.getLaunchTimeoutMillis(), TimeUnit.MILLISECONDS);
+                                LOGGER.log(Level.INFO, String.format("%dms elapsed waiting for node %s to connect", System.currentTimeMillis() - startTime, node.getNodeName()));
+                                return null;
+                            } else {
+                                LOGGER.log(Level.INFO, String.format("No computer for node %s found", node.getNodeName()));
                             }
-                            return node;
```

Comment:
This method still needs to return the node in the success case so it can be added to the list.

Reply:
Actually we are adding this node ourself before even starting its provisioning..

And if its one shot node its sometimes added after being removed..

Reply:
But Actually Im not sure what we should do here.. for sure we dont need to return anything here..
Cause we are doing all logic ourself in plugin..

Reply:
In that case, then I think it might be worthwhile to just return an empty list, rather than add null values.

Reply:
If we will return empty list it will break all other stuff :(
Cause lets assume node is created 30 seconds (thats ~ time for linux, windows is bit slower), and to make quick response for creating node we are running node checking every 10 seconds , then if we will return empty list it will assume that no nodes was created and will create that nodes over and over again..

If we would like to be 100% `NodeProvisioner` compatible we should create nodes ourself before adding them to Jenkins, and most our code now assumes that node is added and then connected...
That mean we would have to rewrite half of plugin :(

Reply:
`NodeProvisioner` is doing only 2 operations with returned node..
- `jenkins.addNode(node);` - we have done this looong time ago, and it fails if node is already deleted.. which might happen when system is really busy..
- `fireOn*(f, node);` - we are not using it.. maybe someone is.. that might be a problem..

Reply:
Understood, thanks for clearing that up. In that case could you please add a TODO comment linking to this issue: https://issues.jenkins-ci.org/browse/JENKINS-55412
Ideally we should be following the normal provisioning workflow and not doing things out of band.

Reply:
So we're manually adding the node [here](https://github.com/jenkinsci/google-compute-engine-plugin/pull/13/files/c60051850cce03f14759411eb73f88ef0f47bb42?utf8=%E2%9C%93&diff=split#diff-13ec31119ed683508c3eaf94108b0ea9R171) which sounds like [an anti-pattern](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/slaves/Cloud.java#L155-L163). I'm confused, though, because we manually added the node _and_ returned a `PlannedNode` in the previous version and we didn't see double nodes. Maybe Jenkins reconciles that?

@ingwarsw you said "most our code now assumes that node is added and then connected...". If we move the [instance provisioning](https://github.com/jenkinsci/google-compute-engine-plugin/pull/13/files/c60051850cce03f14759411eb73f88ef0f47bb42?utf8=%E2%9C%93&diff=split#diff-13ec31119ed683508c3eaf94108b0ea9L159) and don't add the node to Jenkins ourself, where else are we assuming the node is added and connected?

Reply:
@evandbrown Yup.. I suppose its anti-pattern.. 
But I must say that most similar plugins do that..

And jenkins will not add same node twice.. and it was never problem before..
But with this PR node is deleted just after job is finished, and NodeProvisioner is scheduled job, so if job finish/fail really fast it might happen (and It happened a lot in our server) that node is deleted after finishing job, but because its returned from PlannedNode feature its added again.

And regarding "most our code now assumes that node is added and then connected.."
We are using jenkins method to launch instances.. and they all assume that node is added..
Our first line `node.toComputer();` assumes that node is added.. cause computer is created by jenkins when adding node.

To make it perfectly right we should.
Move most of code from launchers to some other classes.. That classes should finish after running jenkins agent..
In launches there should be just setting channel.
And then we should here launch instance from new class, return node and it would be added by NodeProvisioner..

But it have one big drawback..
Now from the start we see than node is provisioned and we can see logs of provisioning in UI.

Reply:
Kubernetes seems to do that right..
https://github.com/jenkinsci/kubernetes-plugin/tree/master/src/main/java/org/csanchez/jenkins/plugins/kubernetes